packages/website/src/components/hooks/useHashState.ts

```diff
@@ -189,7 +188,6 @@ const writeStateToLocalStorage = (newState: ConfigModel): void => {
 function useHashState(
   initialState: ConfigModel,
 ): [ConfigModel, (cfg: Partial<ConfigModel>) => void] {
-  const history = useHistory();
```

Comment:
I'm not quite I understand sure how this fixes the infinite rendering.
Based on everything I'm seeing us using this hook shouldn't be any different to us not using the hook.

Do you have any more information about what components are rerendering?

Reply:
entire playground is re rendered when page is changed, you can actually see page "blinking" if you use visual editor

Reply:
@bradzacher 

> The docusaurus's PendingNavigation component rendered infinitely when the url hash changed using useHistory.
This causes the resize handle to not work as described in the https://github.com/typescript-eslint/typescript-eslint/issues/8731.

It was was inaccurate.

I added a `Debug` component to the `App` and tried to run it. From the logging, it appears that the `DocusaurusContextProvider` is rendering infinitely. 

```tsx
// node_modules/@docusaurus/core/lib/client/App.js


const Debug = ({parent}) => {
  console.log(`[Debug][${parent}]`,'render')
  return <></>
}

export default function App() {
    const routeElement = renderRoutes(routes);
    const location = useLocation();
    return (<ErrorBoundary>
    <Debug parent="ErrorBoundary"/>  {/** <-- Here */}
      <DocusaurusContextProvider>
        <Debug parent="DocusaurusContextProvider"/>  {/** <-- Here */}
        <BrowserContextProvider>
          <Debug parent="BrowserContextProvider"/>
          <Root>
            <Debug parent="Root"/> {/** <-- Here */}
            <SiteMetadataDefaults />
            <SiteMetadata />
            <BaseUrlIssueBanner />
            <PendingNavigation location={normalizeLocation(location)}>
              <Debug parent="PendingNavigation"/>  {/** <-- Here */}
              {routeElement}
            </PendingNavigation>
          </Root>
          <HasHydratedDataAttribute />
        </BrowserContextProvider>
      </DocusaurusContextProvider>
    </ErrorBoundary>);
}

```

https://github.com/typescript-eslint/typescript-eslint/assets/41323220/a0612675-0fdd-4369-b1ee-69eb5b3d51cc

Reply:
It seems to me like this is the wrong fix then and the docusaurus upgrade uncovered a hidden bug in our code.

Realistically the URL hash should be checked on page load to initialise the state - and then all future changes discarded as we rely on internal state updates from that point on.

I.e. We're missing a memoization boundary in our code to prevent hash changes from causing a rerender.

Reply:
that could potentially save some ticks yes, but underlying issue is that query parameter changes trigger now full page reload

Reply:
@bradzacher @armano2 
I set up a simplified reproducible project. https://github.com/yeonjuan/docusaurus-bug-repro
The problem seems to be caused by using `useHistory` inside a `<BrowserOnly />`.

If the following component lazy loaded within `BrowserOnly` (same as Playground), infinite re-renders will be reproduced.

- src/components/Component.jsx
```ts
import { useHistory } from "@docusaurus/router"
import { useCallback, useEffect, useState } from "react";


export default function Component() {

  // bug
  const history = useHistory();
  useEffect(() => {
    history.replace({hash: "hash"});
  }, [])

  return <div>
    LazyLoaded
  </div>
}
```

However, the code below will not cause infinite re-renders when used outside of `<BrowserOnly />`.
```ts
  // bug
  const history = useHistory();
  useEffect(() => {
    history.replace({hash: "hash"});
  }, [])

```

Reply:
I thought it was a bug in docusaurus and reported the issue. (https://github.com/facebook/docusaurus/issues/9976)

Reply:
Oh so the problem here is remount thrashing.

That fix is to define that `lazy` in the module root.

What's happening here is that each time the route changes the `BrowserOnly` component rerenders ( a. It's probably not a `React.memo`/PureComponent and b. We're  passing a function as it's child which is always a new reference - invalidating the props).

This then causes the child function to rerender.
This runs `Component = lazy(..)` again.
We then render Component.

The issue is that `lazy` will return a brand new component each time it runs (it's not a hook like `useMemo` that is stable and render-aware).

So each render react is essentially seeing a brand new component. This causes react to unmount the old tree and mount the new tree because as far as react is concerned that's what we've told it to do.

This is why there's a UI flash and everything.

So yeah the fix is to hoist that `Component = lazy(..)` to the module root so it's stable across renders.
Its okay to do this because the point of `lazy` is that it doesn't call the callback and trigger the imported module load until the component is first rendered.

Reply:
Cool. I've confirmed that it works fine in my machine!!  7e6a077